plugins/producer_plugin/include/eosio/producer_plugin/production_pause_vote_tracker.hpp

```diff
@@ -201,20 +215,27 @@ class production_pause_vote_tracker {
       }
 
       // Can be called concurrently with all member functions.
-      bool record_received_block(fc::time_point block_received_time, fc::microseconds block_acceptance_delta) {
+      bool record_received_block(fc::time_point block_received_time, fc::time_point block_time_stamp, fc::microseconds block_acceptance_delta) {
          auto vote_time = latest_vote.load(std::memory_order_relaxed);
 
-         const auto adjusted_block_received_time = fc::time_point(block_received_time).safe_add(block_acceptance_delta);
+         const auto received_time = std::min(block_received_time, block_time_stamp);
+         const auto adjusted_block_received_time = fc::time_point(received_time).safe_add(block_acceptance_delta);
 
+         // votes (vote_time) can be signaled before the block the vote is associated with is
+         // signaled (received_time). If block received within the block_acceptance_delta tolerance
+         // then act as if the received block came before the last vote.
          if (adjusted_block_received_time <= vote_time)
             return false;
 
          auto block_orig = first_block_after_vote.load(std::memory_order_relaxed);
 
-         if ((vote_time < block_orig) && (block_orig <= block_received_time))
+         // if we have already received a block since our last vote then nothing to do.
+         // Also make sure the block received time is greater than the tracked block, should always
+         // be true unless a clock skew.
+         if ((vote_time < block_orig) && (block_orig <= received_time))
```

Comment:
I added this check before because I wanted to allow `record_received_block` to be thread-safe.

That means two threads might call `record_received_block` block concurrently and so the `store` into the `first_block_after_vote` could come in any order regardless of how their received times were ordered.

But now with the most recent changes of using `received_time = std::min(block_received_time, block_time_stamp)`, there is another reason why they can come out of order.

It is possible that we first accept a block on one branch that has timestamp `t2` and then accept a block on another branch that has timestamp `t1` where `t1 < t2`. Even though the `block_received_time` of the block with timestamp `t1` would presumably be greater than the `block_received_time` of the block with timestamp `t2` (since it was accepted after in nodeos), with the RT being the minimum and both blocks being late, the `received_time` for the first block would be `t2` and the `received_time` for the second block would be the lower value `t1`.

Reply:
Currently `record_received_block` is only called from the main thread. Both `accept_block_header` and `accept_block` are currently only signaled from the main thread. I would like to potentially change that for #529. So maybe in the future it could be called from multiple threads.

I added your comment to the code.